common/src/main/java/com/wynntils/core/net/UrlManager.java

```diff
@@ -61,50 +98,62 @@
+    private static final String OVERRIDE_LINK = System.getProperty(URLS_OVERRIDE_ENV);
+    private final URI urlListOverrideUri;
+
     // -Dwynntils.url.list.ignore.cache=true
     private static final String URLS_IGNORE_CACHE_ENV = "wynntils.url.list.ignore.cache";
+    private static final boolean IGNORE_CACHE = Boolean.parseBoolean(System.getProperty(URLS_IGNORE_CACHE_ENV));
 
-    private final URI urlListOverrideUri;
-    private final boolean ignoreCache;
+    // -Dwynntils.url.list.debug.log=true
+    private static final String URLS_DEBUG_LOG_ENV = "wynntils.url.list.debug.log";
+    private static final boolean DEBUG_LOGS = System.getProperty(URLS_DEBUG_LOG_ENV) != null;
+
+    // -Dwynntils.url.list.force.source=local
+    private static final String URLS_OVERRIDE_FORCE_SOURCE_ENV = "wynntils.url.list.force.source";
+    private static final String FORCE_SOURCE = System.getProperty(URLS_OVERRIDE_FORCE_SOURCE_ENV);
+    private final UrlListType forceSource;
 
-    private Map<UrlId, UrlInfo> urlMap = Map.of();
-    private int version = -1;
+    private Map<UrlListType, UrlList> urlListMap = new ConcurrentHashMap<>();
+
+    private Object urlMapLock = new Object();
+    private UrlList urlList = UrlList.EMPTY;
 
     public UrlManager(NetManager netManager) {
         super(List.of(netManager));
 
         // Read the url override from environment
-        String urlListOverride = System.getProperty(URLS_OVERRIDE_ENV);
-        URI parsedOverrideUri;
-        try {
-            parsedOverrideUri = URI.create(urlListOverride);
-        } catch (Exception e) {
-            parsedOverrideUri = null;
-            WynntilsMod.error("Invalid URL list override link: " + urlListOverride);
+        URI parsedOverrideUri = null;
+        if (OVERRIDE_LINK != null) {
+            try {
+                parsedOverrideUri = URI.create(OVERRIDE_LINK);
+            } catch (Exception e) {
+                parsedOverrideUri = null;
+                WynntilsMod.error("Invalid URL list override link: " + OVERRIDE_LINK);
+            }
         }
         urlListOverrideUri = parsedOverrideUri;
 
-        // Read the ignore cache from environment
-        ignoreCache = Boolean.parseBoolean(System.getProperty(URLS_IGNORE_CACHE_ENV));
-
         // Log the settings
         if (urlListOverrideUri == null) {
             WynntilsMod.info(
-                    "Loading urls.json in the normal mode. Url cache is " + (ignoreCache ? "ignored" : "used") + ".");
+                    "Loading urls.json in the normal mode. Url cache is " + (IGNORE_CACHE ? "ignored" : "used") + ".");
         } else {
             WynntilsMod.info("Loading urls.json from " + urlListOverrideUri + ". Url cache is "
-                    + (ignoreCache ? "ignored" : "used") + ".");
+                    + (IGNORE_CACHE ? "ignored" : "used") + ".");
         }
 
+        forceSource = UrlListType.valueOf(FORCE_SOURCE.toUpperCase(Locale.ROOT));
```

Comment:
This is really putting the spotlight on the need of a SystemPropertiesManager, that can read and initialize values retrieved from system properties, and then just expose those values in their preferred form.

I'll accept it for now, but I'd really appreciate if you could try to make it a priority to refactor out these things afterwards.

Reply:
I'll try my best in the upcoming days.